### PR TITLE
Fix for PR #4595 - Color attachments indices are incorrectly managed

### PR DESCRIPTION
**Pull request by arno-lunarg**
I was forgetting an indirection level, since `vkCmdClearAttachments` refers to sub pass attachments, not render pass attachments.
This PR also adds a corresponding test case.

### CODE REVIEW

File: tests/vklayertests_dynamic_rendering.cpp [hunk 1]
```diff
@@ -379,20 +379,38 @@ TEST_F(VkLayerTest, DynamicRenderingClearAttachments) {
...
 
-    std::array<VkAttachmentReference, 2> attachment_references = {
-        {{0, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL}, {1, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL}}};
-
-    VkSubpassDescription subpass_desc{};
-    subpass_desc.pipelineBindPoint = VK_PIPELINE_BIND_POINT_GRAPHICS;
-    subpass_desc.colorAttachmentCount = 1;
-    subpass_desc.pColorAttachments = &attachment_references[1];
-    subpass_desc.pDepthStencilAttachment = &attachment_references[0];
+    std::vector<VkAttachmentReference> attachment_references = {{{0, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL},
```
Comment (juan-lunarg) [inline tests/vklayertests_dynamic_rendering.cpp:382]: Why is this a vector now?

File: layers/buffer_validation.cpp [hunk 2]
```diff
@@ -3660,13 +3660,19 @@ bool CoreChecks::PreCallValidateCmdClearAttachments(VkCommandBuffer commandBuffe
...
                 const auto *framebuffer = cb_node->activeFramebuffer.get();
 
-                is_valid_color_attachment_index = (clear_desc->colorAttachment != VK_ATTACHMENT_UNUSED) &&
-                                                  (clear_desc->colorAttachment < renderpass_create_info->attachmentCount);
-                if (framebuffer && (clear_desc->colorAttachment != VK_ATTACHMENT_UNUSED) &&
-                    (clear_desc->colorAttachment < framebuffer->createInfo.attachmentCount)) {
-                    color_view_state = cb_node->GetActiveAttachmentImageViewState(clear_desc->colorAttachment);
+                is_valid_color_attachment_index = (clear_desc->colorAttachment == VK_ATTACHMENT_UNUSED);
+                if (subpass_desc) {
+                    is_valid_color_attachment_index |= clear_desc->colorAttachment < subpass_desc->colorAttachmentCount;
```
Comment (Tony-LunarG) [inline layers/buffer_validation.cpp:3665]: Nit - 3 consecutive checks for subpass_desc - combine?  Or don't.

**Review by Tony-LunarG — APPROVED** [review]
LGTM and should get rid of the 700+ new 07270 errors in CTS last night

File: layers/buffer_validation.cpp (hunk 2 above)
Comment (ncesario-lunarg) [reply layers/buffer_validation.cpp:3665]: ➕

File: tests/vklayertests_dynamic_rendering.cpp [hunk 3]
```diff
@@ -417,7 +435,7 @@ TEST_F(VkLayerTest, DynamicRenderingClearAttachments) {
     std::fill(&renderpass_clear_values[0].color.float32[0], &renderpass_clear_values[0].color.float32[0] + 4, 0.0f);
     renderpass_bi.pClearValues = renderpass_clear_values.data();
 
-    auto clear_cmds_func = [&](const bool use_dynamic_rendering) {
+    auto ClearCmdsTest = [&](const bool use_dynamic_rendering) {
```
Comment (ncesario-lunarg) [inline tests/vklayertests_dynamic_rendering.cpp:438]: "technically" this is a variable, and so should remain snake_case (which the [style guide](https://google.github.io/styleguide/cppguide.html#Formatting_Lambda_Expressions) seems to backup) for consistency.

**Review by ncesario-lunarg — APPROVED** [review]
Minor style nit, but otherwise LGTM!

**Comment by arno-lunarg** [comment]
`VkPositiveLayerTest.ExternalFenceSyncFdLoop` failed on the Pixel 6. Should I be worried?

**Comment by jeremyg-lunarg** [comment]
> `VkPositiveLayerTest.ExternalFenceSyncFdLoop` failed on the Pixel 6. Should I be worried?

That's almost certainly not a problem with this PR. I'll take a look at what happened.

**Comment by jeremyg-lunarg** [comment]
> > `VkPositiveLayerTest.ExternalFenceSyncFdLoop` failed on the Pixel 6. Should I be worried?
> 
> That's almost certainly not a problem with this PR. I'll take a look at what happened.

Looks like a vkQueueWaitIdle() call timed out.  Not a problem with this PR. I'll investigate it separately.
```
[ RUN      ] VkPositiveLayerTest.ExternalFenceSyncFdLoop
jni/../../tests/positive/sync.cpp:1714: Failure
Expected equality of these values:
  VK_SUCCESS
    Which is: 0
  resolved_err
    Which is: 2
VK_TIMEOUT
Unexpected: Validation Error: [ VUID-vkDestroyFence-fence-01120 ] Object 0: handle = 0xe7f79a0000000005, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x5d296248 | VkFence 0xe7f79a0000000005[] is in use. The Vulkan spec states: All queue submission commands that refer to fence must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyFence-fence-01120)
Unexpected: Validation Error: [ VUID-vkDestroyFence-fence-01120 ] Object 0: handle = 0xf56c9b0000000004, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x5d296248 | VkFence 0xf56c9b0000000004[] is in use. The Vulkan spec states: All queue submission commands that refer to fence must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyFence-fence-01120)
Unexpected: Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x73de430c40, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xe7f79a0000000005, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x73de430c40[], VkFence 0xe7f79a0000000005[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
Unexpected: Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x73de430c40, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xf56c9b0000000004, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x73de430c40[], VkFence 0xf56c9b0000000004[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkDestroyDevice-device-00378)
[  FAILED  ] VkPositiveLayerTest.ExternalFenceSyncFdLoop (893 ms)
```

**Review by jeremyg-lunarg — APPROVED** [review]

**Pull request merged**